server/proxy/pf_filters.c

```diff
@@ -0,0 +1,218 @@
+
+static const char* FILTER_RESULT_STRINGS[] =
+{
+	"FILTER_PASS",
+	"FILTER_DROP",
+	"FILTER_IGNORE",
+};
+
+static const char* EVENT_TYPE_STRINGS[] =
+{
+	"KEYBOARD_EVENT",
+	"MOUSE_EVENT",
+};
+
+static const char* pf_filters_get_filter_result_string(PF_FILTER_RESULT result)
+{
+	if (result >= FILTER_PASS && result <= FILTER_IGNORE)
+		return FILTER_RESULT_STRINGS[result];
+	else
+		return "FILTER_UNKNOWN";
+}
+
+static const char* pf_filters_get_event_type_string(PF_FILTER_TYPE result)
+{
+	if (result >= FILTER_TYPE_KEYBOARD && result <= FILTER_TYPE_MOUSE)
+		return EVENT_TYPE_STRINGS[result];
+	else
+		return "EVENT_UNKNOWN";
+}
+
+BOOL pf_filters_init(filters_list** list)
+{
+	if (list == NULL)
+	{
+		WLog_ERR(TAG, "pf_filters_init(): list == NULL");
+		return FALSE;
+	}
+
+	*list = ArrayList_New(FALSE);
+
+	if (*list == NULL)
+	{
+		WLog_ERR(TAG, "pf_filters_init(): ArrayList_New failed!");
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+PF_FILTER_RESULT pf_filters_run_by_type(filters_list* list, PF_FILTER_TYPE type,
+                                        connectionInfo* info,
+                                        void* param)
+{
+	proxyFilter* filter;
+	proxyEvents* events;
+	PF_FILTER_RESULT result = FILTER_PASS;
+	const size_t count = (size_t) ArrayList_Count(list);
+	size_t index;
+
+	for (index = 0; index < count; index++)
+	{
+		filter = (proxyFilter*) ArrayList_GetItem(list, index);
+		events = filter->events;
+		WLog_DBG(TAG, "pf_filters_run_by_type(): Running filter: %s", filter->name);
+
+		switch (type)
+		{
+			case FILTER_TYPE_KEYBOARD:
+				IFCALLRET(events->KeyboardEvent, result, info, param);
+				break;
+
+			case FILTER_TYPE_MOUSE:
+				IFCALLRET(events->MouseEvent, result, info, param);
+				break;
+		}
+
+		if (result != FILTER_PASS)
+		{
+			/* Filter returned FILTER_DROP or FILTER_IGNORE. There's no need to call next filters. */
+			WLog_INFO(TAG, "Filter %s [%s] returned %s", filter->name,
+			          pf_filters_get_event_type_string(type), pf_filters_get_filter_result_string(result));
+			return result;
+		}
+	}
+
+	/* all filters returned FILTER_PASS */
+	return FILTER_PASS;
+}
+
+static void pf_filters_filter_free(proxyFilter* filter)
+{
+	assert(filter != NULL);
+	FreeLibrary(filter->handle);
+	free(filter->name);
+	free(filter->events);
+	free(filter);
+}
+
+void pf_filters_unregister_all(filters_list* list)
+{
+	if (list == NULL)
+		return;
+	
+	const size_t count = (size_t) ArrayList_Count(list);
+	size_t index;
+
+	for (index = 0; index < count; index++)
+	{
+		proxyFilter* filter = (proxyFilter*) ArrayList_GetItem(list, index);
+		WLog_DBG(TAG, "pf_filters_unregister_all(): freeing filter: %s", filter->name);
+		pf_filters_filter_free(filter);
+	}
+
+	ArrayList_Free(list);
+}
+
+BOOL pf_filters_register_new(filters_list* list, const char* module_path, const char* filter_name)
+{
+	proxyEvents* events = NULL;
+	proxyFilter* filter = NULL;
+	HMODULE handle = NULL;
+	filterInitFn fn;
+
+	if (list == NULL)
```

Comment:
I don't see what could lead to have `list` being `NULL`

Reply:
changed that check to assert
might happen if a developer calls this function with filters_list uninitialized